src/interface/swarm.cpp

```diff
@@ -317,28 +354,537 @@ void Swarm::Defrag() {
+              bdvar.send[m](buffer_index) = vreal(i, sidx);
+              buffer_index++;
+            }
+            for (int i = 0; i < int_vars_size; i++) {
+              bdvar.send[m](buffer_index) = static_cast<Real>(vint(i, sidx));
+              buffer_index++;
+            }
+            // If rank is shared, apply boundary conditions here
+            // TODO(BRR) Don't hardcode periodic boundary conditions
+            if (nrank(m) == swarm_d.GetMyRank()) {
+              double &x = vreal(ix, sidx);
+              double &y = vreal(iy, sidx);
+              double &z = vreal(iz, sidx);
+              if (x < swarm_d.x_min_global_) {
+                x = swarm_d.x_max_global_ - (swarm_d.x_min_global_ - x);
+              }
+              if (x > swarm_d.x_max_global_) {
+                x = swarm_d.x_min_global_ + (x - swarm_d.x_max_global_);
+              }
+              if (y < swarm_d.y_min_global_) {
+                y = swarm_d.y_max_global_ - (swarm_d.y_min_global_ - y);
+              }
+              if (y > swarm_d.y_max_global_) {
+                y = swarm_d.y_min_global_ + (y - swarm_d.y_max_global_);
+              }
+              if (z < swarm_d.z_min_global_) {
+                z = swarm_d.z_max_global_ - (swarm_d.z_min_global_ - z);
+              }
+              if (z > swarm_d.z_max_global_) {
+                z = swarm_d.z_min_global_ + (z - swarm_d.z_max_global_);
+              }
+            }
+          }
+        }
+      });
+
+  // Count all the particles that are Active and Not on this block, if nonzero,
+  // copy into buffers (if no send already for that buffer) and send
+
+  RemoveMarkedParticles();
+
+  vbswarm->Send(phase);
+  return true;
+}
+
+template <typename T>
+vpack_types::SwarmVarList<T> Swarm::MakeVarListAll_(ParticleVariableVector<T> variables) {
+  int size = 0;
+  vpack_types::SwarmVarList<T> vars;
+
+  for (auto it = variables.rbegin(); it != variables.rend(); ++it) {
+    auto v = *it;
+    vars.push_front(v);
+    size++;
+  }
+  return vars;
+}
+
+SwarmVariablePack<Real> Swarm::PackAllVariablesReal(PackIndexMap &vmap) {
+  std::vector<std::string> names;
+  for (auto &v : realVector_) {
+    names.push_back(v->label());
+  }
+  return PackVariablesReal(names, vmap);
+}
+
+SwarmVariablePack<int> Swarm::PackAllVariablesInt(PackIndexMap &vmap) {
+  std::vector<std::string> names;
+  for (auto &v : intVector_) {
+    names.push_back(v->label());
+  }
+  return PackVariablesInt(names, vmap);
+}
+
+SwarmVariablePack<Real> Swarm::PackVariablesReal(const std::vector<std::string> &names,
+                                                 PackIndexMap &vmap) {
+  vpack_types::SwarmVarList<Real> vars = MakeVarListAll_<Real>(realVector_);
+
+  auto pack = MakeSwarmPack<Real>(vars, &vmap);
+  SwarmPackIndxPair<Real> value;
+  value.pack = pack;
+  value.map = vmap;
+  return pack;
+}
+SwarmVariablePack<int> Swarm::PackVariablesInt(const std::vector<std::string> &names,
```

Comment:
Can these methods be made into a template?

Reply:
I tried to do that but I wasn't able to make that work with the specific references to the `realVector_` and `intVector_` data members. Do you see a workaround?

Reply:
I can try my hand at it.

Reply:
https://github.com/lanl/parthenon/pull/459 here @brryan can you see if this will help you move forward.